### PR TITLE
Implement Atomic write adapter

### PR DESCRIPTION
**Pull request by PatrickRose**
See #567 

This is a "first go" - I'm not entirely sure it's the best way of doing this, and I'm not entirely confident I've added tests in the right place. Feedback greatly appreciated :-)

### CODE REVIEW

**Comment by frankdejonge** [comment]
@PatrickRose This looks like a good first go, I'll have a more thorough look over this in the upcoming week, but it looks like I'll merge this and improve on this if needed.

File: tests/AtomicLocalAdapterTests.php [hunk 1]
```diff
@@ -0,0 +1,356 @@
+<?php
+
+namespace League\Flysystem\Adapter;
+
+require_once('LocalAdapterTests.php');
+
+function rename ($oldname, $newname, $context = null) {
```
Comment (GrahamCampbell) [inline tests/AtomicLocalAdapterTests.php]: cs

File: tests/AtomicLocalAdapterTests.php [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    if (strpos($newname, 'rename.fail') !== false)
+    {
+        return false;
+    }
+
+    return call_user_func_array('rename', func_get_args());
+}
+
+use League\Flysystem\Config;
```
Comment (GrahamCampbell) [inline tests/AtomicLocalAdapterTests.php]: Normally, we put all imports first.

**Comment by PatrickRose** [comment]
@frankdejonge I'd appreciate a ping if you do update this - I'm currently using a version of this for a current project so any improvements would be great (though I'll likely just delete ours and update our deps) plus it's useful to learn :-)

**Pull request closed without merging**